File: src/models/Transaction.ts
import { OrderExecutionType, OrderType, TransactionData } from '../types/stock';

/**
 * Object-Oriented Transaction Model
 * Represents an executed trade record for ledger auditing.
 */
export class Transaction {
  private _id: string;
  private _timestamp: number;
  private _formattedDate: string;
  private _symbol: string;
  private _stockName: string;
  private _type: OrderType;
  private _executionType: OrderExecutionType;
  private _shares: number;
  private _price: number;
  private _totalAmount: number;
  private _fees: number;
  private _realizedPnL?: number;

  constructor(params: {
    id?: string;
    timestamp?: number;
    formattedDate?: string;
    symbol: string;
    stockName: string;
    type: OrderType;
    executionType?: OrderExecutionType;
    shares: number;
    price: number;
    fees?: number;
    realizedPnL?: number;
  }) {
    const ts = params.timestamp || Date.now();
    this._id = params.id || `TX-${ts}-${Math.random().toString(36).substring(2, 7).toUpperCase()}`;
    this._timestamp = ts;
    this._formattedDate = params.formattedDate || new Date(ts).toLocaleString();
    this._symbol = params.symbol.toUpperCase();
    this._stockName = params.stockName;
    this._type = params.type;
    this._executionType = params.executionType || 'MARKET';
    this._shares = params.shares;
    this._price = params.price;
    this._fees = params.fees || 0;
    this._totalAmount = Number((params.shares * params.price + (params.type === 'BUY' ? this._fees : -this._fees)).toFixed(2));
    this._realizedPnL = params.realizedPnL !== undefined ? Number(params.realizedPnL.toFixed(2)) : undefined;
  }

  public get id(): string { return this._id; }
  public get timestamp(): number { return this._timestamp; }
  public get formattedDate(): string { return this._formattedDate; }
  public get symbol(): string { return this._symbol; }
  public get stockName(): string { return this._stockName; }
  public get type(): OrderType { return this._type; }
  public get executionType(): OrderExecutionType { return this._executionType; }
  public get shares(): number { return this._shares; }
  public get price(): number { return this._price; }
  public get totalAmount(): number { return this._totalAmount; }
  public get fees(): number { return this._fees; }
  public get realizedPnL(): number | undefined { return this._realizedPnL; }

  public toJSON(): TransactionData {
    return {
      id: this._id,
      timestamp: this._timestamp,
      formattedDate: this._formattedDate,
      symbol: this._symbol,
      stockName: this._stockName,
      type: this._type,
      executionType: this._executionType,
      shares: this._shares,
      price: this._price,
      totalAmount: this._totalAmount,
      fees: this._fees,
      realizedPnL: this._realizedPnL,
    };
  }
}

File: Transaction.java
package com.trading.model;

import java.io.Serializable;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.UUID;

public class Transaction implements Serializable {
    private static final long serialVersionUID = 1L;

    public enum Type { BUY, SELL, DIVIDEND }
    public enum ExecutionType { MARKET, LIMIT }

    private final String id;
    private final LocalDateTime timestamp;
    private final String symbol;
    private final String stockName;
    private final Type type;
    private final ExecutionType executionType;
    private final int shares;
    private final double price;
    private final double totalAmount;
    private final double fees;
    private final Double realizedPnL;

    public Transaction(String symbol, String stockName, Type type, ExecutionType executionType,
                       int shares, double price, double fees, Double realizedPnL) {
        this.id = "TX-" + UUID.randomUUID().toString().substring(0, 8).toUpperCase();
        this.timestamp = LocalDateTime.now();
        this.symbol = symbol.toUpperCase();
        this.stockName = stockName;
        this.type = type;
        this.executionType = executionType;
        this.shares = shares;
        this.price = price;
        this.fees = fees;
        this.totalAmount = (shares * price) + (type == Type.BUY ? fees : -fees);
        this.realizedPnL = realizedPnL;
    }

    public String getId() { return id; }
    public LocalDateTime getTimestamp() { return timestamp; }
    public String getFormattedDate() {
        return timestamp.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
    }
    public String getSymbol() { return symbol; }
    public String getStockName() { return stockName; }
    public Type getType() { return type; }
    public ExecutionType getExecutionType() { return executionType; }
    public int getShares() { return shares; }
    public double getPrice() { return price; }
    public double getTotalAmount() { return totalAmount; }
    public double getFees() { return fees; }
    public Double getRealizedPnL() { return realizedPnL; }

    @Override
    public String toString() {
        String pnlStr = (realizedPnL != null) ? String.format(" [Realized P&L: %+.2f]", realizedPnL) : "";
        return String.format("[%s] %s %d %s @ $%.2f | Total: $%.2f%s",
                getFormattedDate(), type, shares, symbol, price, totalAmount, pnlStr);
    }
}
